File: backend/src/middleware/rbac.ts
import { Request, Response, NextFunction } from 'express';
import { sendResponse } from '../utils/response.js';

/**
 * Role-based access control middleware.
 * Must be used AFTER `authenticate` so that `req.user` is populated.
 */
export function checkRole(allowedRoles: string[]) {
  return (req: Request, res: Response, next: NextFunction): void => {
    const raw = req.user?.role;
    const roles = Array.isArray(raw)
      ? raw.map((r) => r.toLowerCase())
      : raw
        ? [raw.toLowerCase()]
        : [];

    if (roles.length === 0 || !roles.some((r) => allowedRoles.includes(r))) {
      sendResponse(res, {
        success: false,
        status_code: 403,
        message: 'Không có quyền truy cập',
        message_en: 'Forbidden',
        errors: [`Required roles: ${allowedRoles.join(', ')}`],
      });
      return;
    }

    next();
  };
}

export const isAdmin = checkRole(['admin']);
export const isStaff = checkRole(['admin', 'employee', 'chef']);
export const isChef = checkRole(['admin', 'chef']);
export const isEmployee = checkRole(['admin', 'employee']);
export const isCustomer = checkRole(['customer']);

File: backend/app/middleware/rbac.py
from fastapi import Depends, HTTPException
from app.middleware.auth import authenticate, UserContext

def check_role(allowed_roles: list[str]):
    async def dependency(current_user: UserContext = Depends(authenticate)) -> UserContext:
        user_roles = [r.lower() for r in current_user.role]
        if not any(r in allowed_roles for r in user_roles):
            raise HTTPException(status_code=403, detail="Forbidden")
        return current_user
    return dependency

is_admin = check_role(["admin"])
is_staff = check_role(["admin", "employee", "chef"])
is_chef = check_role(["admin", "chef"])
is_employee = check_role(["admin", "employee"])
is_customer = check_role(["customer", "guest"])
